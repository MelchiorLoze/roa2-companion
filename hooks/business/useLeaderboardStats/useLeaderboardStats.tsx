import { useMemo } from 'react';
import { barDataItem, lineDataItem } from 'react-native-gifted-charts';
import { useUnistyles } from 'react-native-unistyles';

import { useCommunityLeaderboard, useCommunityLeaderboards } from '@/hooks/data';
import { getRank, Rank } from '@/types/rank';

import { useUserStats } from '../useUserStats/useUserStats';

const DISTRIBUTION_PRECISION = 10;

const roundedElo = (elo: number) => Math.floor(elo / DISTRIBUTION_PRECISION) * DISTRIBUTION_PRECISION;

export const useLeaderboardStats = () => {
  const { leaderboards, isLoading: isLoadingLeaderboards } = useCommunityLeaderboards();
  const { leaderboardEntries, isLoading: isLoadingLeaderboard } = useCommunityLeaderboard(leaderboards[1]?.id ?? -1);
  const { stats, isLoading: isLoadingUserStats } = useUserStats();

  const { theme } = useUnistyles();

  const firstPlayerElo = leaderboardEntries[0]?.elo ?? 0;
  const lastPlayerElo = Math.min(0, leaderboardEntries[leaderboardEntries.length - 1]?.elo ?? 0);
  const lastAethereanElo = leaderboardEntries.findLast((entry, index) => entry.elo > 1800 && index < 100)?.elo ?? 1800;

  const distributionPerRank = leaderboardEntries.reduce(
    (acc, entry) => {
      const rank = getRank(entry.elo, entry.position);
      if (acc[rank]?.value) acc[rank].value++;
      else acc[rank] = { value: 1, frontColor: theme.color[rank] };
      return acc;
    },
    {} as Record<Rank, barDataItem>,
  );

  const distributionPerElo = useMemo(() => {
    if (!leaderboardEntries.length || !firstPlayerElo) return [];

    const result = Array.from(
      { length: Math.ceil(firstPlayerElo / DISTRIBUTION_PRECISION) },
      (_, i) => i * DISTRIBUTION_PRECISION,
    ).reduce(
      (acc, elo) => {
        acc[elo] = { value: 0, label: elo.toString() };
        return acc;
      },
      {} as Record<number, lineDataItem>,
    );

    leaderboardEntries.forEach((entry) => {
      const elo = roundedElo(entry.elo);
      if (result[elo].value) result[elo].value++;
      else result[elo].value = 1;
    });

    if (stats?.rankedElo != null) {
      const playerElo = roundedElo(stats.rankedElo);
      if (result[playerElo]) result[playerElo].dataPointColor = 'red';
    }

    return result;
  }, [firstPlayerElo, leaderboardEntries, stats?.rankedElo]);

  return {
    firstPlayerElo,
    lastPlayerElo,
    lastAethereanElo,
    rankDistribution: Object.values(distributionPerRank),
    eloDistribution: Object.values(distributionPerElo),
    isLoading: isLoadingUserStats || isLoadingLeaderboards || isLoadingLeaderboard,
  };
};
