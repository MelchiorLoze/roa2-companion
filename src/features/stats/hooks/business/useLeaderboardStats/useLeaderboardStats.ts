import { useSeason } from '../../../contexts/SeasonContext/SeasonContext';
import { MAX_AETHEREAN_PLAYERS } from '../../../types/rank';
import { useCommunityLeaderboard } from '../../data/useCommunityLeaderboard/useCommunityLeaderboard';

export const useLeaderboardStats = () => {
  const { leaderboardId, isLoading: isLoadingSeason } = useSeason();
  const { leaderboardEntries, isLoading: isLoadingLeaderboard } = useCommunityLeaderboard(leaderboardId);

  const firstPlayerElo = leaderboardEntries[0]?.elo ?? 0;
  const lastPlayerElo = Math.min(0, leaderboardEntries[leaderboardEntries.length - 1]?.elo ?? 0);
  const lastAethereanElo =
    leaderboardEntries.findLast((entry) => entry.elo > 1800 && entry.position < MAX_AETHEREAN_PLAYERS)?.elo ?? 0;

  return {
    firstPlayerElo,
    lastPlayerElo,
    lastAethereanElo,
    leaderboardEntries,
    isLoading: isLoadingSeason || isLoadingLeaderboard,
  };
};
