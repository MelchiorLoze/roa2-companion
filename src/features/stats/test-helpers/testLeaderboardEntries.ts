import { type LeaderboardEntry } from '../types/rank';

export const testLeaderboardEntries: LeaderboardEntry[] = [
  // Aetherean
  { steamId: 1, position: 1, elo: 2162 },
  { steamId: 2, position: 2, elo: 2089 },
  { steamId: 3, position: 3, elo: 1910 },
  { steamId: 4, position: 4, elo: 1837 },
  // Grandmaster
  { steamId: 5, position: 5, elo: 1791 },
  // Master
  { steamId: 6, position: 6, elo: 1655 },
  { steamId: 7, position: 7, elo: 1649 },
  { steamId: 8, position: 8, elo: 1523 },
  // Diamond
  { steamId: 9, position: 9, elo: 1450 },
  { steamId: 10, position: 10, elo: 1300 },
  // Platinum
  { steamId: 11, position: 11, elo: 1258 },
  { steamId: 12, position: 12, elo: 1111 },
  // Gold
  { steamId: 13, position: 13, elo: 1000 },
  { steamId: 14, position: 14, elo: 920 },
  // Silver
  { steamId: 15, position: 15, elo: 820 },
  { steamId: 16, position: 16, elo: 752 },
  // Bronze
  { steamId: 17, position: 17, elo: 699 },
  { steamId: 18, position: 18, elo: 510 },
  // Stone
  { steamId: 19, position: 19, elo: 400 },
  { steamId: 20, position: 20, elo: 300 },
  { steamId: 25, position: 25, elo: -100 },
];
