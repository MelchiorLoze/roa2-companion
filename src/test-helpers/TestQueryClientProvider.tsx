import { QueryClient, type QueryClientConfig, QueryClientProvider } from '@tanstack/react-query';
import { type PropsWithChildren } from 'react';

const queryClientTestConfig: QueryClientConfig = {
  defaultOptions: {
    queries: {
      retry: false,
      gcTime: 0,
      staleTime: 0,
    },
    mutations: {
      retry: false,
      gcTime: 0,
    },
  },
};

export const TestQueryClientProvider = ({ children }: PropsWithChildren) => {
  const queryClient = new QueryClient(queryClientTestConfig);

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};
