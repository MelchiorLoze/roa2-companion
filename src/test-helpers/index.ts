import { Category, type Item, Rarity } from '@/types/item';

export { createItemDto } from './createItemDto';
export { TestQueryClientProvider } from './TestQueryClientProvider';

export const testItemList: Item[] = [
  {
    id: '1',
    name: 'Item 1',
    rarity: Rarity.COMMON,
    category: Category.ICON,
    coinPrice: 2000,
    buckPrice: 20,
  },
  {
    id: '2',
    name: 'Item 2',
    rarity: Rarity.EPIC,
    category: Category.DEATHEFFECT,
    buckPrice: 500,
  },
  {
    id: '3',
    name: 'Item 3',
    rarity: Rarity.RARE,
    category: Category.EMOTE,
    coinPrice: 1500,
  },
  {
    id: '4',
    name: 'Item 4',
    rarity: Rarity.LEGENDARY,
    category: Category.SKIN,
  },
  {
    id: '5',
    name: 'Item 5',
    rarity: Rarity.COMMON,
    category: Category.STAGESKIN,
    coinPrice: 200000,
    buckPrice: 2000,
  },
];
