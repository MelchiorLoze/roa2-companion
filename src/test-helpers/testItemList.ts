import { Category, type Item, Rarity } from '@/types/item';

export const testItemList: Item[] = [
  {
    id: '1',
    friendlyId: 'Icon_item_1',
    name: 'Item 1',
    rarity: Rarity.COMMON,
    category: Category.ICON,
    coinPrice: 2000,
    buckPrice: 20,
  },
  {
    id: '2',
    friendlyId: 'DeathEffect_item_2',
    name: 'Item 2',
    rarity: Rarity.EPIC,
    category: Category.DEATHEFFECT,
    buckPrice: 500,
  },
  {
    id: '3',
    friendlyId: 'Emote_item_3',
    name: 'Item 3',
    rarity: Rarity.RARE,
    category: Category.EMOTE,
    coinPrice: 1500,
  },
  {
    id: '4',
    friendlyId: 'Skin_item_4',
    name: 'Item 4',
    rarity: Rarity.LEGENDARY,
    category: Category.SKIN,
  },
  {
    id: '5',
    friendlyId: 'StageSkin_item_5',
    name: 'Item 5',
    rarity: Rarity.COMMON,
    category: Category.STAGESKIN,
    coinPrice: 200000,
    buckPrice: 2000,
  },
];
