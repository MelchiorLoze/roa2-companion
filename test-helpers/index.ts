export { createItemDto } from './createItemDto';
export { TestQueryClientProvider } from './TestQueryClientProvider';
