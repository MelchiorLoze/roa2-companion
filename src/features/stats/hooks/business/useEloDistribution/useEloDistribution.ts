import { type LeaderboardEntry } from '../../../types/rank';
import { useLeaderboardStats } from '../useLeaderboardStats/useLeaderboardStats';

const DISTRIBUTION_PRECISION = 20;

const roundElo = (elo: number) => Math.floor(elo / DISTRIBUTION_PRECISION) * DISTRIBUTION_PRECISION;

const getEloDistribution = (
  leaderboardEntries: LeaderboardEntry[],
  firstPlayerElo: number,
  lastPlayerElo: number,
): Record<number, number> => {
  if (!leaderboardEntries.length) return {};

  const totalEloRange = firstPlayerElo - lastPlayerElo + 1;
  const result = Array.from(
    { length: Math.ceil(totalEloRange / DISTRIBUTION_PRECISION) },
    (_, i) => i * DISTRIBUTION_PRECISION + roundElo(lastPlayerElo),
  ).reduce((acc, elo) => ({ ...acc, [elo]: 0 }), {} as Record<number, number>);

  leaderboardEntries.forEach((entry) => result[roundElo(entry.elo)]++);

  return result;
};

export const useEloDistribution = () => {
  const { firstPlayerElo, lastPlayerElo, leaderboardEntries, isLoading } = useLeaderboardStats();

  return {
    eloDistribution: getEloDistribution(leaderboardEntries, firstPlayerElo, lastPlayerElo),
    roundElo,
    isLoading,
  };
};
