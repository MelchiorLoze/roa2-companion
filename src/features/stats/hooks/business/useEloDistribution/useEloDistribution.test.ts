import { renderHook } from '@testing-library/react-native';

import { testLeaderboardEntries } from '../../../test-helpers/testLeaderboardEntries';
import { useLeaderboardStats } from '../useLeaderboardStats/useLeaderboardStats';
import { useEloDistribution } from './useEloDistribution';

jest.mock('../useLeaderboardStats/useLeaderboardStats');
const useLeaderboardStatsMock = jest.mocked(useLeaderboardStats);
const defaultLeaderboardStatsState = {
  firstPlayerElo: 2162,
  lastPlayerElo: -100,
  lastAethereanElo: 1837,
  leaderboardEntries: testLeaderboardEntries,
  isLoading: false,
};

const renderUseEloDistribution = () => {
  const { result } = renderHook(useEloDistribution);

  expect(useLeaderboardStatsMock).toHaveBeenCalledTimes(1);

  return { result };
};

describe('useEloDistribution', () => {
  beforeEach(() => {
    useLeaderboardStatsMock.mockReturnValue(defaultLeaderboardStatsState);
  });

  it('returns correct values when leaderboard stats are loading', () => {
    useLeaderboardStatsMock.mockReturnValue({
      ...defaultLeaderboardStatsState,
      firstPlayerElo: 0,
      lastPlayerElo: 0,
      lastAethereanElo: 0,
      leaderboardEntries: [],
      isLoading: true,
    });

    const { result } = renderUseEloDistribution();

    expect(result.current.isLoading).toBe(true);
    expect(result.current.eloDistribution).toEqual({});
  });

  it('returns correct elo distribution for complete leaderboard', () => {
    const { result } = renderUseEloDistribution();

    expect(result.current.isLoading).toBe(false);
    expect(result.current.eloDistribution).toMatchSnapshot();
  });
});
