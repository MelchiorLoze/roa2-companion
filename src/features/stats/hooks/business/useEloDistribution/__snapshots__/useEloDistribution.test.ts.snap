// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useEloDistribution returns correct elo distribution for complete leaderboard 1`] = `
{
  "-100": 1,
  "-20": 0,
  "-40": 0,
  "-60": 0,
  "-80": 0,
  "0": 0,
  "100": 0,
  "1000": 1,
  "1020": 0,
  "1040": 0,
  "1060": 0,
  "1080": 0,
  "1100": 1,
  "1120": 0,
  "1140": 0,
  "1160": 0,
  "1180": 0,
  "120": 0,
  "1200": 0,
  "1220": 0,
  "1240": 1,
  "1260": 0,
  "1280": 0,
  "1300": 1,
  "1320": 0,
  "1340": 0,
  "1360": 0,
  "1380": 0,
  "140": 0,
  "1400": 0,
  "1420": 0,
  "1440": 1,
  "1460": 0,
  "1480": 0,
  "1500": 0,
  "1520": 1,
  "1540": 0,
  "1560": 0,
  "1580": 0,
  "160": 0,
  "1600": 0,
  "1620": 0,
  "1640": 2,
  "1660": 0,
  "1680": 0,
  "1700": 0,
  "1720": 0,
  "1740": 0,
  "1760": 0,
  "1780": 1,
  "180": 0,
  "1800": 0,
  "1820": 1,
  "1840": 0,
  "1860": 0,
  "1880": 0,
  "1900": 1,
  "1920": 0,
  "1940": 0,
  "1960": 0,
  "1980": 0,
  "20": 0,
  "200": 0,
  "2000": 0,
  "2020": 0,
  "2040": 0,
  "2060": 0,
  "2080": 1,
  "2100": 0,
  "2120": 0,
  "2140": 0,
  "2160": 1,
  "220": 0,
  "240": 0,
  "260": 0,
  "280": 0,
  "300": 1,
  "320": 0,
  "340": 0,
  "360": 0,
  "380": 0,
  "40": 0,
  "400": 1,
  "420": 0,
  "440": 0,
  "460": 0,
  "480": 0,
  "500": 1,
  "520": 0,
  "540": 0,
  "560": 0,
  "580": 0,
  "60": 0,
  "600": 0,
  "620": 0,
  "640": 0,
  "660": 0,
  "680": 1,
  "700": 0,
  "720": 0,
  "740": 1,
  "760": 0,
  "780": 0,
  "80": 0,
  "800": 0,
  "820": 1,
  "840": 0,
  "860": 0,
  "880": 0,
  "900": 0,
  "920": 1,
  "940": 0,
  "960": 0,
  "980": 0,
}
`;
